boards/arm/stm32/nucleo-f4x1re/src/stm32_ssd1306.c

```diff
@@ -0,0 +1,164 @@
+
+static struct lcd_dev_s    *g_lcddev;
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: board_ssd1306_initialize
+ *
+ * Description:
+ *   Initialize and register the device. I2C version.
+ *
+ * Input Parameters:
+ *   busno - The I2C bus number
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_LCD_SSD1306_I2C
+int board_ssd1306_initialize(int busno)
+{
+  struct i2c_master_s *i2c;
+  const int devno = 0;
+
+  /* Initialize I2C */
+
+  i2c = stm32_i2cbus_initialize(busno);
+  if (!i2c)
+    {
+      lcderr("ERROR: Failed to initialize I2C port %d\n", busno);
+      return -ENODEV;
+    }
+
+  /* Bind the I2C port to the OLED */
+
+  g_lcddev = ssd1306_initialize(i2c, NULL, devno);
+  if (!g_lcddev)
+    {
+      lcderr("ERROR: Failed to bind I2C port %d to OLED %d\n", busno, devno);
+      return -ENODEV;
+    }
+  else
+    {
+      lcdinfo("Bound I2C port %d to OLED %d\n", busno, devno);
+
+      /* And turn the OLED on */
+
+      g_lcddev->setpower(g_lcddev, CONFIG_LCD_MAXPOWER);
+      return OK;
+    }
+}
+#endif
+
+/****************************************************************************
+ * Name: board_ssd1306_initialize
+ *
+ * Description:
+ *   Initialize and register the device. SPI version.
+ *
+ * Input Parameters:
+ *   busno - The SPI bus number
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_LCD_SSD1306_SPI
+int board_ssd1306_initialize(int busno)
+{
+  struct spi_dev_s *spi;
+  const int devno = 0;
+
+  /* Initialize SPI */
+
+  spi = stm32_spibus_initialize(busno);
+  if (!spi)
+    {
+      lcderr("ERROR: Failed to initialize SPI port %d\n", busno);
+      return -ENODEV;
+    }
+
+  /* Bind the SPI port to the OLED */
+
+  g_lcddev = ssd1306_initialize(spi, NULL, devno);
+  if (!g_lcddev)
+    {
+      lcderr("ERROR: Failed to bind SPI port %d to OLED %d\n", busno, devno);
+      return -ENODEV;
+    }
+  else
+    {
+      lcdinfo("Bound SPI port %d to OLED %d\n", busno, devno);
+
+      /* And turn the OLED on */
+
+      g_lcddev->setpower(g_lcddev, CONFIG_LCD_MAXPOWER);
+      return OK;
+    }
+}
+#endif
+
+/****************************************************************************
+ * Name: board_ssd1306_getdev
+ *
+ * Description:
+ *   Get the SSD1306 device driver instance
+ *
+ * Returned Value:
+ *   Pointer to the instance
+ *
+ ****************************************************************************/
+
+struct lcd_dev_s *board_ssd1306_getdev(void)
```

Comment:
Maybe registry can be done at caller level? I mean that `board_ssd1306_initialize` and `board_ssd1306_initialize` car return `struct lcd_dev_s *` or `NULL` and `board_lcd_initialize` can store the pointer to LCD device.

Reply:
The source of the problem is the fact, that I had to copy file stm32_ssd1306.c from directory board/arm/stm32/**common**/src to directory board/arm/stm32/**nucleo-f4x1re**/src in order to solve linker problem **undefined reference to `board_ssd1306_getdev**. I will try to fix the linker problem without copying the file **stm32_ssd1306.c** to the board/src directory. In case of redundant file (stm32_ssd1306.c in board/src directory) registry could be done at caller level without major changes.